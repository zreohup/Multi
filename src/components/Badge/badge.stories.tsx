import type { Meta, StoryObj } from '@storybook/react'
import { Badge } from '@/src/components/Badge'
import { SafeFontIcon } from '@/src/components/SafeFontIcon/SafeFontIcon'
import React from 'react'
import { Text, View } from 'tamagui'

const meta: Meta<typeof Badge> = {
  title: 'Badge',
  component: Badge,
  args: {
    content: '3/9',
  },
}

export default meta

type Story = StoryObj<typeof Badge>

export const Circular: Story = {
  args: {
    content: '12+',
  },
}
export const CircularWithIcon: Story = {
  render: function Render(args) {
    return <Badge {...args} content={<SafeFontIcon size={13} name="owners" />} />
  },
}
export const NonCircular: Story = {
  args: {
    content: 'Badge',
    circular: false,
  },
}

export const NonCircularBold: Story = {
  args: {
    content: 'Badge',
    circular: false,
    textContentProps: {
      fontWeight: 700,
    },
  },
}

export const NonCircularWithComplexContent: Story = {
  args: {
    circular: false,
  },
  render: function Render(args) {
    return (
      <Badge
        {...args}
        content={
          <View alignItems="center" flexDirection="row" gap="$1">
            <SafeFontIcon size={13} name="owners" />

            <Text fontWeight={600} color={'$color'}>
              3/9
            </Text>
          </View>
        }
      />
    )
  },
}
