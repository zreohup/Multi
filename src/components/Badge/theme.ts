import { tokens } from '@/src/theme/tokens'

export const badgeTheme = {
  light_badge_success: {
    background: tokens.color.successLightDark,
    color: tokens.color.backgroundMainDark,
  },
  dark_badge_success: {
    color: tokens.color.backgroundMainDark,
    background: tokens.color.backgroundLightDark,
  },
  light_badge_success_variant1: {
    background: tokens.color.successBackgroundLight,
    color: tokens.color.successMainLight,
  },
  dark_badge_success_variant1: {
    background: tokens.color.successBackgroundDark,
    color: tokens.color.successMainLight,
  },
  light_badge_success_variant2: {
    background: tokens.color.secondaryLightLight,
    color: tokens.color.primaryMainLight,
  },
  dark_badge_success_variant2: {
    background: tokens.color.secondaryMainLight,
    color: tokens.color.primaryMainLight,
  },
  light_badge_warning: {
    color: tokens.color.warning1ContrastTextLight,
    background: tokens.color.warningBackgroundLight,
  },
  dark_badge_warning: {
    color: tokens.color.warning1ContrastTextDark,
    background: tokens.color.warningBackgroundDark,
  },
  light_badge_warning_variant2: {
    color: tokens.color.warning1TextLight,
    background: tokens.color.warning1ContrastTextLight,
  },
  dark_badge_warning_variant2: {
    color: tokens.color.warning1MainDark,
    background: tokens.color.warning1ContrastTextDark,
  },
  dark_badge_background: {
    color: tokens.color.textPrimaryDark,
    background: tokens.color.backgroundSecondaryDark,
  },
  light_badge_background: {
    color: tokens.color.textPrimaryLight,
    background: tokens.color.backgroundSecondaryLight,
    borderColor: tokens.color.logoBackgroundLight,
  },
  light_badge_error: {
    color: tokens.color.errorMainLight,
    background: tokens.color.errorBackgroundLight,
  },
  dark_badge_error: {
    color: tokens.color.errorMainDark,
    background: tokens.color.errorDarkDark,
  },
  light_badge_background_inverted: {
    color: tokens.color.logoBackgroundLight,
    background: tokens.color.textPrimaryLight,
  },
  dark_badge_background_inverted: {
    color: tokens.color.logoBackgroundDark,
    background: tokens.color.textPrimaryDark,
  },
}
