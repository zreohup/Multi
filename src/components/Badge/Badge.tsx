import React from 'react'
import { Circle, CircleProps, SizeTokens, Text, TextProps, Theme, View } from 'tamagui'
import { badgeTheme } from '@/src/components/Badge/theme'

type BadgeThemeKeys = keyof typeof badgeTheme

type ExtractAfterUnderscore<T extends string> = T extends `${string}_${infer Rest}` ? Rest : never
export type BadgeThemeTypes = ExtractAfterUnderscore<BadgeThemeKeys>

interface BadgeProps {
  content: string | React.ReactElement
  themeName?: BadgeThemeTypes
  circleSize?: string | SizeTokens
  fontSize?: TextProps['fontSize']
  circleProps?: Partial<CircleProps>
  textContentProps?: Partial<TextProps>
  circular?: boolean
  testID?: string
}

export const Badge = ({
  content,
  circleSize = '$7',
  fontSize = 14,
  themeName = 'badge_warning',
  circular = true,
  circleProps,
  textContentProps,
  testID,
}: BadgeProps) => {
  let contentToRender = content
  if (typeof content === 'string') {
    contentToRender = (
      <Text fontSize={fontSize} color={'$color'} {...textContentProps}>
        {content}
      </Text>
    )
  }

  if (circular) {
    return (
      <Theme name={themeName}>
        <Circle
          testID={testID}
          size={circleSize}
          backgroundColor={'$background'}
          borderColor={'$borderColor'}
          {...circleProps}
        >
          {contentToRender}
        </Circle>
      </Theme>
    )
  }
  return (
    <Theme name={themeName}>
      <View
        testID={testID}
        alignSelf={'flex-start'}
        paddingVertical="$1"
        paddingHorizontal="$3"
        gap="$1"
        borderRadius={50}
        backgroundColor={'$background'}
        {...circleProps}
      >
        {contentToRender}
      </View>
    </Theme>
  )
}
