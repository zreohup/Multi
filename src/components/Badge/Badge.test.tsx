import React from 'react'
import { render } from '@/src/tests/test-utils'
import { Badge } from './Badge'

describe('Badge', () => {
  it('renders circular badge', () => {
    const view = render(<Badge content="12+" testID="badge" />)
    expect(view).toMatchSnapshot()
  })

  it('renders non-circular badge', () => {
    const view = render(<Badge content="Label" circular={false} testID="badge" />)
    expect(view).toMatchSnapshot()
  })

  it('renders badge with custom size and theme', () => {
    const view = render(<Badge content="ok" circleSize="$8" themeName="badge_success" testID="badge" />)
    expect(view).toMatchSnapshot()
  })
})
