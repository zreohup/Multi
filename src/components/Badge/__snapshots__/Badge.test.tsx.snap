// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders badge with custom size and theme 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#3B7A54",
        "borderBottomLeftRadius": 100000,
        "borderBottomRightRadius": 100000,
        "borderTopLeftRadius": 100000,
        "borderTopRightRadius": 100000,
        "flexDirection": "column",
        "height": 32,
        "justifyContent": "center",
        "maxHeight": 32,
        "maxWidth": 32,
        "minHeight": 32,
        "minWidth": 32,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
        "width": 32,
      }
    }
    testID="badge"
  >
    <Text
      style={
        {
          "color": "#121312",
          "fontFamily": "DM Sans",
          "fontSize": 14,
        }
      }
      suppressHighlighting={true}
    >
      ok
    </Text>
  </View>
</View>
`;

exports[`Badge renders circular badge 1`] = `
<View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFECC2",
        "borderBottomLeftRadius": 100000,
        "borderBottomRightRadius": 100000,
        "borderTopLeftRadius": 100000,
        "borderTopRightRadius": 100000,
        "flexDirection": "column",
        "height": 28,
        "justifyContent": "center",
        "maxHeight": 28,
        "maxWidth": 28,
        "minHeight": 28,
        "minWidth": 28,
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
        "width": 28,
      }
    }
    testID="badge"
  >
    <Text
      style={
        {
          "color": "#FF8C00",
          "fontFamily": "DM Sans",
          "fontSize": 14,
        }
      }
      suppressHighlighting={true}
    >
      12+
    </Text>
  </View>
</View>
`;

exports[`Badge renders non-circular badge 1`] = `
<View>
  <View
    style={
      {
        "alignSelf": "flex-start",
        "backgroundColor": "#FFECC2",
        "borderBottomLeftRadius": 50,
        "borderBottomRightRadius": 50,
        "borderTopLeftRadius": 50,
        "borderTopRightRadius": 50,
        "gap": 4,
        "paddingBottom": 4,
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingTop": 4,
      }
    }
    testID="badge"
  >
    <Text
      style={
        {
          "color": "#FF8C00",
          "fontFamily": "DM Sans",
          "fontSize": 14,
        }
      }
      suppressHighlighting={true}
    >
      Label
    </Text>
  </View>
</View>
`;
